20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,124 @@
+
+The specific technology/product used for the KMS can be decided in a later stage. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+#### Satellite Project Updates
+
+A new flag should be added to the `projects` table, like `projects.satellite_managed_encryption`. This should be false for all existing projects. While satellite-managed-encryption (this design) is in-progress, new projects should be created with this flag set to false.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the new column in the `projects` table is set to `true` for this project
+* a new key-value pair is created for the new project in the Key Management Store. The passphrase should be cryptographically random
+* endpoints on the satellite API are added or updated so that an authenticated user can request the encryption passphrase for a project during access creation
+
+#### WebAssembly Updates
+
+Wasm functionality must be updated to allow the Satellite UI to properly generate access credentials with path encryption disabled. [Code for reference](https://github.com/storj/storj/blob/a62929fd5757a40eda5d0b044ad2cefc14708410/satellite/console/consolewasm/access.go#L22-L29).
+
+#### Satellite UI Updates
+
+When creating a project, the user should be provided the option to create different types of projects:
+
+* satellite-managed-encryption project
+* user-managed-encryption project
+
+The user-managed-encryption project will be the same as before, and user will always be prompted to enter encryption passphrase when necessary, e.g. when using the file browser or generating an access grant.
+
+The satellite-managed-encryption project will never prompt the user for an encryption passphrase in the UI. Instead, when the user does an action that requires generating credentials, the UI will retrieve the project's encryption passphrase from the satellite KMS, and generate an access grant using this passphrase, and path encryption disabled.
+
+The user should not be able to directly retrieve a "raw API key" from the Satellite UI. This prevents the user from unintentionally creating "invalid" credentials using a raw API key with Uplink. The only credentials that can be generated from the access page for a satellite-managed-encryption project should be credentials that already have embedded encryption rules.
+
+The file browser in the UI should generate access credentials using the passphrase from the satellite KMS, and no path encryption.
+
+#### Uplink Updates
```

Comment:
Yes it could, but I don't think these changes are completely necessary given the other changes described. If the UI only provides the user with a restricted access credential, it will be difficult for the user to accidentally setup uplink with the wrong encryption, even if we make no changes

Reply:
I meant, that we should mention that those would require updates as we are mentioning here that uplink-cli does.